htlcswitch/link_test.go

```diff
@@ -807,7 +800,115 @@ func TestExitNodeAmountPayloadMismatch(t *testing.T) {
+// incoming HTLC, if the amount encoded in the onion payload of the forwarded
+// HTLC exceeds the incoming payment value, then the HTLC will be rejected.
+func TestExitNodeAmountPayloadExceedsHTLC(t *testing.T) {
+	t.Parallel()
+
+	f := func(underpayment_rand uint64) bool {
+		underpayment := lnwire.MilliSatoshi(
+			underpayment_rand % (btcutil.SatoshiPerBitcoin - 1) + 1,
+		)
+		channels, _, err := createClusterChannels(
+			t, btcutil.SatoshiPerBitcoin*5,
+			btcutil.SatoshiPerBitcoin*5,
+		)
+		require.NoError(t, err, "unable to create channel")
+
+		n := newThreeHopNetwork(t, channels.aliceToBob,
+			channels.bobToAlice, channels.bobToCarol,
+			channels.carolToBob, testStartingHeight)
+		if err := n.start(); err != nil {
+			t.Fatal(err)
+		}
+		t.Cleanup(n.stop)
+
+		const amount = btcutil.SatoshiPerBitcoin
+		htlcAmt, htlcExpiry, hops := generateHops(amount,
+			testStartingHeight, n.firstBobChannelLink)
+
+		// In order to exercise this case, we'll now _manually_ modify
+		// the per-hop payload for amount to be the incorrect value.
+		// The acceptable values of the amount to forward should be less
+		// than the incoming HTLC value.
+		hops[0].FwdInfo.AmountToForward = amount + underpayment
+		firstHop := n.firstBobChannelLink.ShortChanID()
+		_, err = makePayment(
+			n.aliceServer, n.bobServer, firstHop, hops, amount,
+			htlcAmt, htlcExpiry,
+		).Wait(30 * time.Second)
+		if err == nil {
+			t.Fatalf("payment should have failed but didn't")
+		}
+		assertFailureCode(t, err, lnwire.CodeFinalIncorrectHtlcAmount)
+		return err != nil
+	}
+	if err:= quick.Check(f, &quick.Config{MaxCount: 20}); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// TestExitNodeHTLCExceedsAmountPayload tests that when an exit node receives an
+// incoming HTLC, if the amount encoded in the onion payload of the forwarded
+// HTLC is lower than the incoming HTLC value, then the HTLC will be accepted.
+func TestExitNodeHTLCExceedsAmountPayload(t *testing.T) {
+	t.Parallel()
+
+	f := func(overpayment_rand uint64) bool {
+		overpayment := lnwire.MilliSatoshi(
+			overpayment_rand % (btcutil.SatoshiPerBitcoin - 1) + 1,
+		)
+		channels, _, err := createClusterChannels(
+			t, btcutil.SatoshiPerBitcoin*5,
+			btcutil.SatoshiPerBitcoin*5,
+		)
+		require.NoError(t, err, "unable to create channel")
+
+		n := newThreeHopNetwork(t, channels.aliceToBob,
+			channels.bobToAlice, channels.bobToCarol,
+			channels.carolToBob, testStartingHeight)
+		if err := n.start(); err != nil {
+			t.Fatal(err)
+		}
+		t.Cleanup(n.stop)
+
+		const amount = btcutil.SatoshiPerBitcoin
+		htlcAmt, htlcExpiry, hops := generateHops(amount,
+			testStartingHeight, n.firstBobChannelLink)
+
+		// In order to exercise this case, we'll now _manually_ modify
+		// the per-hop payload for amount to be the incorrect value.
+		// The acceptable values of the amount to forward should be
+		// lower than the incoming HTLC value.
+		hops[0].FwdInfo.AmountToForward = amount - overpayment
```

Comment:
Is there a check somewhere against the original invoice amount?  What happens if both the payload `AmountToForward` and the final hop HTLC amount are below the invoice amount?  Bob *should* fail in that case, but does he?

Reply:
I believe that is beyond the scope of the changes being made. The logic that I am touching is solely around the final hop onion payload integrity checks. The question I believe you are asking here is how we handle HTLCs that don't cover the full amount of the invoice which is the responsibility of the existing invoice code.

Reply:
> Is there a check somewhere against the original invoice amount? What happens if both the payload AmountToForward and the final hop HTLC amount are below the invoice amount? Bob should fail in that case, but does he?

Yes, this is in another sub-system though, found here: https://github.com/lightningnetwork/lnd/blob/86e7b4e1e02e29117dcebba319b0da98e66a3365/invoices/update.go#L196-L200

`inv` is the invoice read from disk. 

The link mainly ensures that the content in the `update_add_htlc` message (created by the incoming node) is congruent with the info in the onion payload (created by the sender).